Add support for readonly array types ('readonly x[]')

mjambon opened the pull request:
This adds support for the special syntax for read-only arrays. It's not much code and the tests pass... but please note the following:

* I removed the right-precedence annotation because I don't see how or when it would apply. I may be missing something.
* I specified a precedence of `PREC.ARRAY_TYPE+1` so as to make `readonly` less sticky. Not sure if declaring a name for the new precedence value would be better.

mjambon commented:
ping :)

maxbrunsfeld commented:
From the [handbook](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#readonly-tuples), It looks like the `readonly` modifier is also supported on tuple types (e.g. `readonly [string, boolean`]). Can you incorporate that into this PR?

mjambon commented:
I just added support for readonly tuple types. I also added support for read-only index signatures ([example](https://github.com/chef/automate/blob/ea100bb080730b3ef08f081c84211abc6c38900f/components/automate-ui/src/app/entities/entities.ts#L20) reported at https://github.com/returntocorp/semgrep/issues/1713).

maxbrunsfeld commented:
Looks great, thanks!
